plugins/router/f5/f5.go

```diff
@@ -0,0 +1,555 @@
+package f5
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/golang/glog"
+)
+
+type F5Error struct {
+	verb           string
+	url            string
+	httpStatusCode int
+	err            error
+	JsonStatusCode int     `json:"code"`
+	Message        *string `json:"message"`
+}
+
+func (err F5Error) Error() string {
+	var msg string
+
+	if err.err != nil {
+		msg = fmt.Sprintf("error: %v", err.err)
+	} else if err.Message != nil {
+		msg = fmt.Sprintf("HTTP code: %d; error from F5: %s",
+			err.httpStatusCode, *err.Message)
+	} else {
+		msg = fmt.Sprintf("HTTP code: %d.", err.httpStatusCode)
+	}
+
+	return fmt.Sprintf("Encountered an error on %s request to URL %s: %s",
+		err.verb, err.url, msg)
+}
+
+type f5LTM struct {
+	host     string
+	username string
+	password string
+	vserver  string
+	insecure bool
+
+	poolMembers map[string]map[string]bool
+}
+
+type f5LTMCfg struct {
+	host     string
+	username string
+	password string
+	vserver  string
+	insecure bool
+}
+
+const policy = "openshift_application_aliases"
+
+func newF5LTM(cfg f5LTMCfg) (*f5LTM, error) {
+	router := &f5LTM{
+		host:        cfg.host,
+		username:    cfg.username,
+		password:    cfg.password,
+		vserver:     cfg.vserver,
+		insecure:    cfg.insecure,
+		poolMembers: map[string]map[string]bool{},
+	}
+
+	return router, nil
+}
+
+//
+// Helper routines for REST calls.
+//
+
+// One of three things can happen as a result of a REST request to F5
+// iControl:
+//
+// (1) The request succeeds and F5 returns an HTTP 200 response, possibly with
+//     a result payload, which should have the fields defined in the
+//     result argument.  In this case, rest_request decodes the payload into
+//     the result argument and returns nil.
+//
+// (2) The request fails and F5 returns an HTTP 4xx or 5xx response with a
+//     response payload containing a code (which should be the same as the
+//     HTTP response code) and a string message.  In this case, rest_request
+//     decodes the response payload and returns an F5Error with the URL, HTTP
+//     verb, HTTP status code, and error information from the response payload.
+//
+// (3) The REST call fails in some other way, such as a socket error or an
+//     error decoding the result payload.  In this case, rest_request returns
+//     an F5Error with the URL, HTTP verb, HTTP status code (if any), and error
+//     value.
+func (f5 *f5LTM) rest_request(verb string, url string, payload io.Reader,
+	result interface{}) error {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: f5.insecure},
+	}
+
+	errorResult := F5Error{verb: verb, url: url}
+
+	req, err := http.NewRequest(verb, url, payload)
+	if err != nil {
+		errorResult.err = fmt.Errorf("http.NewRequest failed: %v", err)
```

Comment:
Yeah, that's what the big fancy `F5Error` structure and its `Error` method give you—that seemed like the golangish way to do it.  Is it failing to log something that it could be logging, or is the approach overengineered or otherwise bad?